### ISSUE TITLE
ansible-lint does not support dynamic `hosts` list (syntax-check error)

### ISSUE THREAD

**Issue by jgspratt** [ansible/ansible-lint#2304]
##### Summary

There seems to be no way to use `ansible-lint` on a playbook which defines its host list dynamically (that is, based on variables).  This technique is important as it allows the reduction in the number of playbooks that need to be written.

##### Ansible and Ansible Lint details

```console
# ansible --version
ansible [core 2.13.3]
  config file = None
  configured module search path = ['/Users/Justin.Spratt/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  ansible collection location = /Users/Justin.Spratt/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.9.9 (main, Nov 21 2021, 03:22:47) [Clang 12.0.0 (clang-1200.0.32.29)]
  jinja version = 3.0.3
  libyaml = True


# ansible-lint --version
WARNING: PATH altered to include /usr/local/opt/python@3.9/bin
ansible-lint 6.4.0 using ansible 2.13.3
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT

Mac OS Monterey (also tested on AWS Linux 2).

##### STEPS TO REPRODUCE

```yaml
---
- name: test
  hosts:
    - "RoleName_{{ role }}"

  become: true
  become_user: root
  gather_facts: false

  tasks:
    - name: test
      debug:
        msg: hello
```

##### Desired Behavior

ansible-lint should not error out.  At the very least, an override should be possible here.

##### Actual Behavior

```
# ansible-lint test.yml
WARNING: PATH altered to include /usr/local/opt/python@3.9/bin
INFO     Set ANSIBLE_LIBRARY=/Users/[redacted]/.cache/ansible-compat/b79606/modules:/Users/[redacted]/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/Users/[redacted]/.cache/ansible-compat/b79606/collections:/Users/[redacted]/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/Users/[redacted]/.cache/ansible-compat/b79606/roles:roles:/Users/[redacted]/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
INFO     Set ANSIBLE_LIBRARY=/Users/[redacted]/.cache/ansible-compat/b79606/modules:/Users/[redacted]/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/Users/[redacted]/.cache/ansible-compat/b79606/collections:/Users/[redacted]/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/Users/[redacted]/.cache/ansible-compat/b79606/roles:roles:/Users/[redacted]/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: test.yml
INFO     Executing syntax check on test.yml (0.87s)
WARNING  Listing 1 violation(s) that are fatal
syntax-check: The field 'hosts' has an invalid value, which includes an undefined variable. The error was: 'role' is undefined
test.yml:2:3 ERROR! The field 'hosts' has an invalid value, which includes an undefined variable. The error was: 'role' is undefined

The error appears to be in '/Users/[redacted]/git/[redacted]/src/ansible/config/test.yml': line 2, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

---
- name: test
  ^ here



Finished with 1 failure(s), 0 warning(s) on 1 files.
```

**Comment by ssbarnea** [comment]
Can you please provide a link to some official documentation about this feature? That would quite useful because there are many things that can be done with Ansible that are not officially supported and we do not want to encourage them, as they are likely to break at any point in the future.

**Comment by jgspratt** [comment]
In Ansible, you can use variables inside `{{ }}`.  You can read about it here: https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html

Specifically, this is the list variables in order of precedence: https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#understanding-variable-precedence

Those are officially supported.  The previous version we were using was `4.2.0` and it did not error out on these lines.  If newer versions will not support all variable types, we will be forced to stop using ansible-lint.

**Comment by ssbarnea** [comment]
Wait a little bit, there is a WIP PR that is rewriting the jinja2 processing that needs to go in first. I will take care of it right after.

**Comment by ssbarnea** [comment]
This is not a bug, it is a coding mistake in your playbook for not providing a default value for the potentially undefined "role" variable. `ansible-playbook --syntax-check` would also fail. It is up to you to ensure you have valid defaults or that you provide defaults in inventory.

Try replacing with `    - "RoleName_{{ 'role' }}"` and you will see you will not get an error. That is because it will template the string role and not an undefined variable.

**Comment by jgspratt** [comment]
This works fine for me.  Give it a try.

```
ansible-playbook --syntax-check -e 'role=foo' test.yml
```

**Comment by ssbarnea** [comment]
I know that it works but you should have have to pass variables to Ansible to be able to perform a syntax check on it. If you need this, it means you have (sorry) poor code. There are multiple ways you can avoid it, like having a default value for the variables defined inside the playbook, checking it presence and using fail action, or just using a `| default` filter approach. How you fix it is up to you, the linter cannot really tell you how.

I will try to update the online documentation of syntax check to explain this issue. You are not the first one facing it and we think it is the time to provide a better explanation.

**Comment by jgspratt** [comment]
Why does passing required variables to Ansible mean that I have poor code?

**Comment by ssbarnea** [comment]
Nobody writes perfect code and very few know about adding defaults for variables. We updated the documentation to include exactly the kind of example you asked about https://ansible-lint.readthedocs.io/en/latest/default_rules/#syntax-check
